src/backends/plonky2/primitives/merkletree/mod.rs

```diff
@@ -82,30 +76,83 @@ impl MerkleTree {
         &mut self,
         key: &RawValue,
         value: &RawValue,
-    ) -> TreeResult<MerkleProofStateTransition> {
+    ) -> TreeResult<MerkleTreeStateTransitionProof> {
         let proof_non_existence = self.prove_nonexistence(key)?;
 
         let old_root: Hash = self.root.hash();
         self.root
-            .add_leaf(0, self.max_depth, Leaf::new(self.max_depth, *key, *value)?)?;
+            .apply_op(self.max_depth, MerkleTreeOp::Insert, *key, Some(*value))?;
         let new_root = self.root.compute_hash();
 
         let (v, proof) = self.prove(key)?;
         assert!(proof.existence);
         assert_eq!(v, *value);
         assert!(proof.other_leaf.is_none());
 
-        Ok(MerkleProofStateTransition {
-            typ: 0, // insertion
+        Ok(MerkleTreeStateTransitionProof {
+            op: MerkleTreeOp::Insert, // insertion
             old_root,
-            proof_non_existence,
+            op_proof: proof_non_existence,
             new_root,
-            new_key: *key,
-            new_value: *value,
+            op_key: *key,
+            op_value: *value,
+            value: None,
             siblings: proof.siblings,
         })
     }
 
+    pub fn update(
+        &mut self,
+        key: &RawValue,
+        value: &RawValue,
+    ) -> TreeResult<MerkleTreeStateTransitionProof> {
+        let (old_value, old_proof) = self.prove(key)?;
+
+        let old_root: Hash = self.root.hash();
+        self.root
+            .apply_op(self.max_depth, MerkleTreeOp::Update, *key, Some(*value))?;
+        let new_root = self.root.compute_hash();
+
+        let (v, proof) = self.prove(key)?;
+        assert!(proof.existence);
+        assert_eq!(v, *value);
+        assert!(proof.other_leaf.is_none());
+
+        Ok(MerkleTreeStateTransitionProof {
+            op: MerkleTreeOp::Update,
+            old_root,
+            op_proof: old_proof,
+            new_root,
+            op_key: *key,
+            op_value: *value,
+            value: Some(old_value),
+            siblings: proof.siblings,
+        })
+    }
+
+    pub fn delete(&mut self, key: &RawValue) -> TreeResult<MerkleTreeStateTransitionProof> {
+        let (value, proof_existence) = self.prove(key)?;
+
+        let old_root: Hash = self.root.hash();
+        self.root
+            .apply_op(self.max_depth, MerkleTreeOp::Delete, *key, None)?;
+        let new_root = self.root.compute_hash();
+
+        let proof = self.prove_nonexistence(key)?;
+        assert!(!proof.existence);
+
+        Ok(MerkleTreeStateTransitionProof {
+            op: MerkleTreeOp::Delete,
+            old_root,
+            op_proof: proof,
+            new_root,
+            op_key: *key,
+            op_value: value,
+            value: None,
+            siblings: proof_existence.siblings,
```

Comment:
In the other operations `op_proof` contains the proof before the operation, and `siblings` the siblings after the operation.  Why is this different?
Then I see that in `verify_state_transition` for `delete` you call `insert` just swapping the `old_root` and `new_root`, so you end up with something that is correct.

Wouldn't it be more consistent if
- `fn delete` returns `op_proof: proof_existence` and `siblings: proof.siblings`
- In `verify_state_transition` for the delete case you rearrange the `op_proof` and `siblings`

Reply:
Update: I just saw that this behavior is documented in the `MerkleTreeStateTransitionProof` type, so I think this is fine :)

Reply:
Update2: After reviewing the circuit implementation I'm even more convinced that this is the way to go.  The circuit treats the siblings/proof "differently" for the delete in order to skip constraints.  By having the native implementation like this we get a native implementation that matches the behavior of the circuits.  I think it's more important that the native implementation matches the circuits than having the consistency I was mentioning in the first message.